### ISSUE TITLE
UML ordered properties

### ISSUE THREAD

**Issue by jakebeal** [SD2E/paml#27]
In UML, some properties have ordered values, which we need to be able to represent correctly in RDF.

An example of this is the `operand` property of an `Expression`, in which the order determines, for example, the difference between "a < b" and "b < a".

RDF triples, however, are inherently unordered:
```
my_expression operand a
my_expression operand b
```
is intentionally equivalent to:
```
my_expression operand b
my_expression operand a
```
so we have to represent the order explicitly, rather than implicitly as is done with UML's serialization format.

There are a number of possible approaches that can be taken here. Here are the candidates that I see, along with their pros and cons:
1. Make an `Ordered` interface, which adds a required index value to an object in the same way that `TypedElement` adds a type to the object, or an `Orderable` interface that adds an optional index.
    - Pros: Succinct, RDF-compatible
    - Cons: Setting an ordering on an _object_ as opposed to a property only works if the object is referred to in precisely one place.
2. Make the whole ordered list the value of the property.
    - Pros: represents "ordered list" explicitly
    - Cons: will break RDF indexing, since a value with URLs embedded in it won't match the individual URLs.
3. Convert ordered properties into named properties. 
    - Pros: This is effectively what is being done in my implementation of Behavior calls right now, in which the name of the InputPin is matched with the name of the Parameter. Also allows keyword argument calls and makes argument alignment explicit rather than implicit.
    - Cons: That might be wrong to do. Also, it would involve modifying the UML class hierarchy.
4. Add an `OrderedPropertyValue` class that serves as a wrapper for any other class, with an `index` field and a `property_value` field.
    - Pros: Succinct, RDF-compatible
    - Cons: Will need to subclass in order to distinguish between compositional and non-compositional references.

Of these options, I think the `OrderedPropertyValue` approach is likely to be the best, since it avoids either modifying the UML class hierarchy or breaking RDF tooling. I believe it is a faithful instantiation of the UML concept that any property can be ordered by making any property subject to an ordering relation. The subclassing required will not be any worse than we already have for things like the different types of literal value, and should be able to be handled implicitly in many cases.

Where it is compatible with the UML, however, I think it would be good to take advantage of the named properties approach instead (e.g., for calling behaviors), since that has positive properties in explicit argument alignment.

**Comment by danbryce** [comment]
`OrderedPropertyValue` seems to make the most sense.

**Comment by jakebeal** [comment]
Great! @danbryce , do you want to take the first pass on this, given that it's directly entangled with your work adding the UML expressions?

**Comment by danbryce** [comment]
Yes, I can do it.  I am close to done with the extensions for time related UML constructs and will add this as a separate PR once the other additions are stable.
